react-forms/src/InputWithButton.tsx

```diff
@@ -0,0 +1,140 @@
+//
+//  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
+//  the NOTICE file distributed with this work for additional information regarding copyright ownership.
+//  The GPA licenses this file to you under the MIT License (MIT), the "License"; you may not use this
+//  file except in compliance with the License. You may obtain a copy of the License at:
+//
+//      http://opensource.org/licenses/MIT
+//
+//  Unless agreed to in writing, the subject software distributed under the License is distributed on an
+//  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
+//  License for the specific language governing permissions and limitations.
+//
+//  Code Modification History:
+//  ----------------------------------------------------------------------------------------------------
+//  0502/2024 - Preston Crawford
+//       Generated original version of source code.
+//
+// ******************************************************************************************************
+
+import * as React from 'react';
+import HelperMessage from './HelperMessage';
+import { CreateGuid, IsInteger, IsNumber } from '@gpa-gemstone/helper-functions'
+
+interface IProps<T> {
+    Record: T;
+    Field: keyof T;
+    Setter: (record: T) => void;
+    Valid: (field: keyof T) => boolean;
+    Label?: string;
+    BtnLabel: string;
+    OnBtnClick: (evt: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
+    BtnClass?: string;
+    Feedback?: string;
+    Disabled?: boolean;
+    Type?: 'number' | 'text' | 'password' | 'email' | 'color' | 'integer';
+    Help?: string | JSX.Element;
+    Style?: React.CSSProperties;
+    AllowNull?: boolean;
+    Size?: 'small' | 'large',
+    DefaultValue?: number,
+    BtnDisabled?: boolean
+}
+
+
+export default function InputWithButton<T>(props: IProps<T>) {
+    const internal = React.useRef<boolean>(false);
+    const [guid, setGuid] = React.useState<string>("");
+    const [showHelp, setShowHelp] = React.useState<boolean>(false);
+    const [heldVal, setHeldVal] = React.useState<string>(''); // Need to buffer tha value because parseFloat will throw away trailing decimals or zeros
+
+    React.useEffect(() => {
+        setGuid(CreateGuid());
+    }, []);
+
+    React.useEffect(() => {
+        if (!internal.current) {
+            setHeldVal(props.Record[props.Field] == null ? '' : (props.Record[props.Field] as any).toString());
+        }
+        internal.current = false;
+    }, [props.Record[props.Field]]);
+
+    function onBlur() {
+        const allowNull = props.AllowNull === undefined ? false : props.AllowNull;
+        if (!allowNull && (props.Type === 'number' || props.Type === 'integer') && heldVal === '') {
+            internal.current = false;
+            props.Setter({ ...props.Record, [props.Field]: props.DefaultValue ?? 0 });
+        }
+    }
+
+    function valueChange(value: string) {
```

Comment:
another small nitpick, this one might not be worth the marginal performance gain for the effort, especially if we memoize the whole thing: these functions re-assign every rerender. By making  a React.useCallback, we can provide a dep array to only reassign when something it depends on changes.

Reply:
I ended up just wrapping the component in a memo.

Reply:
That's not a great solution. `memo` can have unintended consequences.
If we memonize the whole thing it should be done outside of gemstone and we do not want to rely on it every time.